File: leetcode/21 Merge Two Sorted Lists/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        if (list1 == null) return list2; // 左右边界！！！
        if (list2 == null) return list1;        
        ListNode current = new ListNode();
        if (list1.val < list2.val) {
            current = list1;
            current.next = mergeTwoLists(list1.next, list2); // recursion！！！
        } else {
            current = list2;
            current.next = mergeTwoLists(list1, list2.next);          
        }
        return current;
    }
}

File: leetcode/21 Merge Two Sorted Lists/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
    if list1 == nil {  // 左右边界！！！
        return list2
    }
    if list2 == nil {
        return list1
    }
    var current *ListNode = nil
    if list1.Val < list2.Val {
        current = list1
        current.Next = mergeTwoLists(list1.Next, list2)  // recursion！！！
    } else {
        current = list2
        current.Next = mergeTwoLists(list1, list2.Next)
    }
    return current
}
